torchrl/modules/tensordict_module/rnn.py

```diff
@@ -0,0 +1,234 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+from typing import Optional, Tuple
+
+import torch
+
+from tensordict.nn import TensorDictModuleBase as ModuleBase
+
+from tensordict.tensordict import NO_DEFAULT, TensorDictBase
+from tensordict.utils import prod
+
+from torch import nn
+
+
+class LSTMModule(ModuleBase):
+    """An embedder for an LSTM module.
+
+    This class adds the following functionality to :class:`torch.nn.LSTM`:
+
+    - Compatibility with TensorDict: the hidden states are reshaped to match
+      the tensordict batch size.
+    - Optional multi-step execution: with torch.nn, one has to choose between
+      :class:`torch.nn.LSTMCell` and :class:`torch.nn.LSTM`, the former being
+      compatible with single step inputs and the latter being compatible with
+      multi-step. This class enables both usages.
+
+
+    After construction, the module is *not* set in temporal mode, ie. it will
+    expect single steps inputs.
+
+    If in temporal mode, it is expected that the last dimension of the tensordict
+    marks the number of steps. There is no constrain on the dimensionality of the
+    tensordict (except that it must be greater than one for temporal inputs).
+
+    Args:
+        lstm (torch.nn.Module): a :class:`torch.nn.LSTM` instance. For simplicity, this class does not
+            handle the construction of the module.
+        in_keys (list of str): a triplet of strings corresponding to the input value,
+            first and second hidden key.
+        out_keys (list of str): a triplet of strings corresponding to the output value,
+            first and second hidden key.
+            .. note::
+              For a better integration with TorchRL's environments, the best naming
+              for the output hidden key is ``("next", "hidden0")`` or similar, such
+              that the hidden values are passed from step to step during a rollout.
+
+    Attributes:
+        temporal_mode: Returns the temporal mode of the module.
+
+    Methods:
+        set_temporal_mode: controls whether the module should be executed in
+            temporal mode.
+
+    Examples:
+        >>> from torchrl.envs import TransformedEnv, InitTracker
+        >>> from torchrl.envs.libs.gym import GymEnv
+        >>> from torchrl.modules import MLP
+        >>> from torch import nn
+        >>> from tensordict.nn import TensorDictSequential as Seq, TensorDictModule as Mod
+        >>> env = TransformedEnv(GymEnv("Pendulum-v1"), InitTracker())
+        >>> lstm = nn.LSTM(input_size=env.observation_spec["observation"].shape[-1], hidden_size=64, batch_first=True)
+        >>> lstm_module = LSTMModule(lstm, in_keys=["observation", "hidden0", "hidden1"], out_keys=["intermediate", ("next", "hidden0"), ("next", "hidden1")])
```

Comment:
With the comment above this becomes
`LSTMModule(lstm, in_keys=["observation"], out_keys=["intermediate"])`